controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,856 @@
+   // array of clock enable registers, and the bit number used to enable
+   // the clock for the specified peripherial.
+   int ndx = -1;
+   int bit = 0;
+   switch( base )
+   {
+      case DMA1_BASE:    ndx = 0; bit =  0; break;
+      case DMA2_BASE:    ndx = 0; bit =  1; break;
+      case FLASH_BASE:   ndx = 0; bit =  8; break;
+
+      case GPIO_A_BASE:  ndx = 1; bit =  0; break;
+      case GPIO_B_BASE:  ndx = 1; bit =  1; break;
+      case GPIO_C_BASE:  ndx = 1; bit =  2; break;
+      case GPIO_D_BASE:  ndx = 1; bit =  3; break;
+      case GPIO_E_BASE:  ndx = 1; bit =  4; break;
+      case GPIO_H_BASE:  ndx = 1; bit =  7; break;
+      case ADC_BASE:     ndx = 1; bit = 13; break;
+
+      case TIMER2_BASE:  ndx = 4; bit =  0; break;
+      case TIMER3_BASE:  ndx = 4; bit =  1; break;
+      case TIMER6_BASE:  ndx = 4; bit =  4; break;
+      case TIMER7_BASE:  ndx = 4; bit =  7; break;
+      case SPI2_BASE:    ndx = 4; bit = 14; break;
+      case SPI3_BASE:    ndx = 4; bit = 15; break;
+      case UART2_BASE:   ndx = 4; bit = 17; break;
+      case UART3_BASE:   ndx = 4; bit = 18; break;
+      case UART4_BASE:   ndx = 4; bit = 19; break;
+      case I2C1_BASE:    ndx = 4; bit = 21; break;
+      case I2C2_BASE:    ndx = 4; bit = 22; break;
+      case I2C3_BASE:    ndx = 4; bit = 23; break;
+
+      case I2C4_BASE:    ndx = 5; bit =  1; break;
+
+      case TIMER1_BASE:  ndx = 6; bit = 11; break;
+      case SPI1_BASE:    ndx = 6; bit = 12; break;
+      case UART1_BASE:   ndx = 6; bit = 13; break;
+      case TIMER15_BASE: ndx = 6; bit = 16; break;
+      case TIMER16_BASE: ndx = 6; bit = 17; break;
+   }
+
+   // Not sure what to do if the input address isn't found.
+   // For now I just return.  Maybe I should force a fault?
+   if( ndx < 0 ) return;
+
+   // Enable the clock of the requested peripherial
+   RCC_Regs *rcc = (RCC_Regs *)RCC_BASE;
+   rcc->periphClkEna[ndx] |= (1<<bit);
+}
+
+/******************************************************************
+ * Interrupt vector table.  The interrupt vector table is a list of
+ * pointers to the various interrupt functions.  It is stored at the
+ * very start of the flash memory.
+ *****************************************************************/
+
+// Fault handlers
+static void fault( void )
+{
+   while(1){}
+}
+
+static void NMI( void )           { fault(); }
+static void FaultISR( void )      { fault(); }
+static void MPUFaultISR( void )   { fault(); }
+static void BusFaultISR( void )   { fault(); }
+static void UsageFaultISR( void ) { fault(); }
+static void BadISR( void )        { fault(); }
+
+extern "C" void Reset_Handler( void );
+__attribute__ ((section(".isr_vector")))
+void (* const vectors[])(void) =
+{
+   // The first entry of the ISR holds the initial value of the
+   // stack pointer.  The ARM processor initializes the stack
+   // pointer based on this address.
+   (void (*)(void))((uint32_t)mainStack + sizeof(mainStack)),
```

Comment:
Please use C++-style reinterpret_cast instead of the less safe C-style casts throughout.